### DISCUSSION TITLE
dangerouslySetInnerHTML not working with Taillwind and React.js

### DISCUSSION THREAD

**Discussion by AayushBiswas** [tailwindlabs/tailwindcss#7756] [Help] [answered]
I am creating an app and fetching tailwind codes from an localhost file the file Code section looks like this ``` <button class=\"px-4 py-2 rounded-full bg-amber-500 text-white hover:bg-amber-600 transition duration-300 cursor-pointer\">Lorem</button> ``` It's json. Slashes are not counted. but on dangerouslySetInnerHTML it doesn't work

**Comment by LOUTAG** [comment, marked as answer]
Hi, 
I had the same problem that I could fix by adding json file extension to `tailwind.config.js`

> module.exports = {
>   content: [
>     "./src/**/*.{js,jsx,ts,tsx,json}"
>   ],
>   theme: {
>     extend: {},
>   },
>   plugins: [],
> }

**Reply by thenotoriousgustav** [reply]
still doesnt work me, i need a help

**Reply by luis030821** [reply]
the unique solution than find is using cdn of tailwind

**Reply by isaacdarcilla** [reply]
Not working

**Comment by pandalabco** [comment]
dangerouslySetInnerHTML expects to be passed a string of HTML, not a string of JSX.

The class attribute is used directly. You can't adjust the className property instead.

Here is the [stack overflow](https://stackoverflow.com/questions/70837511/why-styles-dont-apply-on-dangerouslysetinnerhtml) where I found the answer

**Comment by henriquematiasb** [comment]
Not sure if this solution would apply to buttons, but working with prismic what worked for me was using `@tailwindcss/typography` plugin adding the `prose` class to work with more sensible styles in your texts as the tailwind documentation points out [here](https://tailwindcss.com/docs/typography-plugin#installation).

This way the code would be:

```<div className="prose" dangerouslySetInnerHTML={{ __html: your_content }} />```

**Reply by isaacdarcilla** [reply]
This works! Thank you @henriquematiasb

**Comment by raqibnur** [comment]
Insert Tailwind css classes inside your css class.
As an example for adding line break in for your content:
`.lineBreak {
  @apply lg:block;
}`

**Comment by Amir-Zouerami** [comment]
This is way past due but in case someone in the future had the same problem read this.

The accepted answer did not work for me so I did the following.

## Locate _WHERE_ you are getting the HTML string from
In my case it was in a file called `utils.ts` inside a directory called `utility`. 

## Edit your `tailwind.config.js` to include the file
Now you can go to `tailwind.config.js`, under the `content` section and make sure your file is added:

```ts
const config: Config = {
  content: [
    './pages/**/*.{js,ts,jsx,tsx,mdx}',
    './components/**/*.{js,ts,jsx,tsx,mdx}',
    './app/**/*.{js,ts,jsx,tsx,mdx}',
    './utility/**/*.{js,ts,jsx,tsx,mdx}', // <--- THIS IS THE LINE I ADDED TO MY CONFIG TO INCLUDE MY FILE
  ],
  darkMode: 'class',
// the rest of your config
```

by the way, `utils.ts` had a function that manipulated the data (HTML string) I was getting from my database before showing it (added some tailwind classes to it and changed some tags). By including this file in the `tailwind.config.js` the problem was solved.

`content` in tailwind config: https://tailwindcss.com/docs/content-configuration

_sidenote: pay attention to your file format (ts, json, etc.)._

**Reply by AzmeerHamasAli** [reply]
This is because the HTML document searches for the "class" tag but our rendered HTML document/component has the tag "className". So HTML considers that the class attribute is missing and doesn't render the tailwind. So to prevent this make sure that where every you are storing you comment as a string instead of "className" and use "class".

**Reply by LeonardoStrechar** [reply]
actually, passing "class=tailwind code" works, I tested with className="tailwind code" and class={{css code}} and it didn't work

**Reply by LeonardoStrechar** [reply]
I tested with Html decode and did not get positive results

**Comment by intelcoder** [comment]
When you try to run `npx tailwindcss -i ./src/index.css -o ./src/ouput.css` do you see all the classes you need in the `ouput.css` ?

Vite or webpack does tree-shaking on the classes so if the classes are not in the project will be ignored. Therefore, you need to add all the classes you need to html string in safelist in the tailwind.config.js

https://tailwindcss.com/docs/content-configuration#safelisting-classes
```
  safelist: [
    'bg-red-500',
    'text-3xl',
    'lg:text-4xl',
  ]
```

**Comment by Dyllan-Betika** [comment]
For anyone who runs into the same issue, I've implemented a working solution specifically for my own requirements...But it's a pretty much boiler plate implementation. 🎯

Read through tailwind typography documentation.
[tailwindcss-typography](https://github.com/tailwindlabs/tailwindcss-typography)

Once you've installed the dependencies, and update your tailwind config file, you should be good to.
Ensure you use the class"prose", which allows you to render out rich text content.

I've also created a helper function to render any "dangerouslySetInnerHTML" content:
`export const createMarkup = (htmlString: string) => {
  return { __html: htmlString };
};`
